opa-manager-operator/src/charm.py

```diff
@@ -0,0 +1,220 @@
+
+
+class CustomResourceDefintion(object):
+    def __init__(self, name, spec):
+
+        self._name = name
+        self._spec = spec
+
+    @property
+    def spec(self):
+        return self._spec
+
+    @property
+    def name(self):
+        return self._name
+
+
+class OPAManagerCharm(CharmBase):
+    """
+    A Juju Charm for OPA
+    """
+
+    _stored = StoredState()
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.update_status, self._on_update_status)
+        self.framework.observe(self.on.stop, self._on_stop)
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.start, self._on_start)
+        self._stored.set_default(things=[])
+
+    def _on_config_changed(self, _):
+        """
+        Set a new Juju pod specification
+        """
+        self._configure_pod()
+
+    def _on_stop(self, _):
+        """
+        Mark unit is inactive
+        """
+        self.unit.status = MaintenanceStatus("Pod is terminating.")
+        logger.info("Pod is terminating.")
+
+    def _load_yaml_objects(self, files_list):
+        yaml_objects = []
+        try:
+            yaml_objects = [yaml.load(Path(f).read_text()) for f in files_list]
+        except yaml.YAMLError as exc:
+            print("Error in configuration file:", exc)
+
+        return yaml_objects
+
+    def _on_install(self, event):
+        logger.info("Congratulations, the charm was properly installed!")
+
+    def _on_update_status(self, event):
+        logger.info("Status updated")
+
+    def _build_pod_spec(self):
+        """
+        Construct a Juju pod specification for OPA
+        """
+        logger.debug("Building Pod Spec")
+
+        # Load Custom Resource Definitions
+        crd_objects = [
+            CustomResourceDefintion(crd["metadata"]["name"], yaml.dump(crd["spec"]))
+            for crd in self._load_yaml_objects(
+                [
+                    "files/configs.config.gatekeeper.sh.yaml",
+                    "files/constrainttemplates.templates.gatekeeper.sh.yaml",
+                    "files/constraintpodstatuses.status.gatekeeper.sh.yaml",
+                    "files/constrainttemplatepodstatuses.status.gatekeeper.sh.yaml",
+                ]
+            )
+        ]
+
+        # Load custom resources
+        crs = [
+            CustomResourceDefintion(cr["metadata"]["name"], yaml.dump(cr))
+            for cr in self._load_yaml_objects(["files/psp.yaml"])
+        ]
+
+        config = self.model.config
+
+        template_args = {
+            "crds": crd_objects,
+            "image_path": config["imagePath"],
+            "imagePullPolicy": config["imagePullPolicy"],
+            "app_name": self.app.name,
+            "cli_args": self._cli_args(),
+            "audit_cli_args": self._audit_cli_args(),
+            "crs": crs,
```

Comment:
Looks like `audit_cli_args` and `crs` aren't referenced in the pod-spec template. Please remove them from the template args.

Reply:
ah yes, I was trying to create them as a single charm - an experiment that failed. removed.